File: sdk/src/policy.mjs
// Write-mode, error-policy, and conflict-resolution enums for the sync client.
//
// These are ENUMS, not booleans, on purpose: each write names its exact
// optimism level and error surface, and new levels can be added without
// breaking call sites. The wire values are canonical for every language port
// (protocol/write-policy.schema.json; the Rust core src/policy.rs and Dart
// lib/src/policy.dart mirror them; zed-interfaces re-exports them).

/** How optimistically a single write is applied. @enum {string} */
export const WriteMode = Object.freeze({
  /** Apply locally + enqueue durably; no network until the next flush. */
  LOCAL_ONLY: "local_only",
  /** Apply locally, enqueue, send; a send failure leaves it queued (default). */
  OPTIMISTIC_QUEUE: "optimistic_queue",
  /** Like OPTIMISTIC_QUEUE, but a send failure is surfaced per the error policy. */
  OPTIMISTIC_AWAIT_ACK: "optimistic_await_ack",
  /** Send first; apply locally only after the ack. */
  SERVER_FIRST: "server_first",
  /** Send only; local updates when the change feed delivers the committed row. */
  SERVER_ONLY: "server_only",
});

/** How write/flush errors are surfaced. @enum {string} */
export const ErrorPolicy = Object.freeze({
  /** Reject the returned promise. */
  THROW_ONLY: "throw_only",
  /** Report through telemetry/onError, resolve normally (default for writes). */
  EMIT_ONLY: "emit_only",
  /** Both. */
  THROW_AND_EMIT: "throw_and_emit",
  /** Neither; the outcome is only visible in the result. */
  SILENT: "silent",
});

/** How a dirty-vs-newer-remote conflict resolves. @enum {string} */
export const ConflictResolution = Object.freeze({
  /** Adopt the server row, drop the queued local write (default). */
  SERVER_WINS: "server_wins",
  /** Keep the higher HLC version — leaderless multi-master converges. */
  LAST_WRITE_WINS: "last_write_wins",
});

const WRITE_MODES = new Set(Object.values(WriteMode));
const ERROR_POLICIES = new Set(Object.values(ErrorPolicy));
const CONFLICTS = new Set(Object.values(ConflictResolution));

/** @param {string} mode */
export function assertWriteMode(mode) {
  if (!WRITE_MODES.has(mode)) {
    throw new RangeError(
      `unknown sync write mode "${mode}" (expected one of ${[...WRITE_MODES].join(", ")})`,
    );
  }
  return mode;
}

/** @param {string} policy */
export function assertErrorPolicy(policy) {
  if (!ERROR_POLICIES.has(policy)) {
    throw new RangeError(
      `unknown sync error policy "${policy}" (expected one of ${[...ERROR_POLICIES].join(", ")})`,
    );
  }
  return policy;
}

/** @param {string} resolution */
export function assertConflictResolution(resolution) {
  if (!CONFLICTS.has(resolution)) {
    throw new RangeError(
      `unknown conflict resolution "${resolution}" (expected one of ${[...CONFLICTS].join(", ")})`,
    );
  }
  return resolution;
}

/** @param {string} policy */
export const shouldThrow = (policy) =>
  policy === ErrorPolicy.THROW_ONLY || policy === ErrorPolicy.THROW_AND_EMIT;
/** @param {string} policy */
export const shouldEmit = (policy) =>
  policy === ErrorPolicy.EMIT_ONLY || policy === ErrorPolicy.THROW_AND_EMIT;

File: dart/zed_sync/lib/src/policy.dart
// Write-mode, error-policy, and conflict-resolution enums — the Dart mirror of
// protocol/write-policy.schema.json (and src/policy.rs, sdk/src/policy.mjs).
// Enums, not booleans: each write names its exact optimism level and error
// surface. `wire` is the canonical cross-language string.

enum WriteMode {
  localOnly('local_only'),
  optimisticQueue('optimistic_queue'),
  optimisticAwaitAck('optimistic_await_ack'),
  serverFirst('server_first'),
  serverOnly('server_only');

  final String wire;
  const WriteMode(this.wire);

  static WriteMode fromWire(String w) =>
      WriteMode.values.firstWhere((m) => m.wire == w,
          orElse: () => throw ArgumentError('unknown write mode "$w"'));
}

enum ErrorPolicy {
  throwOnly('throw_only'),
  emitOnly('emit_only'),
  throwAndEmit('throw_and_emit'),
  silent('silent');

  final String wire;
  const ErrorPolicy(this.wire);

  bool get shouldThrow =>
      this == ErrorPolicy.throwOnly || this == ErrorPolicy.throwAndEmit;
  bool get shouldEmit =>
      this == ErrorPolicy.emitOnly || this == ErrorPolicy.throwAndEmit;

  static ErrorPolicy fromWire(String w) =>
      ErrorPolicy.values.firstWhere((p) => p.wire == w,
          orElse: () => throw ArgumentError('unknown error policy "$w"'));
}

enum ConflictResolution {
  serverWins('server_wins'),
  lastWriteWins('last_write_wins');

  final String wire;
  const ConflictResolution(this.wire);

  static ConflictResolution fromWire(String w) =>
      ConflictResolution.values.firstWhere((c) => c.wire == w,
          orElse: () =>
              throw ArgumentError('unknown conflict resolution "$w"'));
}
